monty-31/src/aarch64_neon/poseidon2.rs

```diff
@@ -93,9 +340,206 @@ where
+        //
+        // This computes the "negative form" of the constant.
+        let input_sub_p = input - (MP::PRIME as i32);
+
+        // Broadcast (duplicate) the scalar result into all four lanes of a 128-bit NEON vector.
+        let vec_s32 = aarch64::vdupq_n_s32(input_sub_p);
+
+        // Reinterpret the bits of the signed vector as an unsigned vector.
+        //
+        // This is a zero-cost operation that only changes the type for the Rust compiler.
+        aarch64::vreinterpretq_u32_s32(vec_s32)
+    }
+}
+
+/// Performs the AddRoundConstant and S-Box operations of a Poseidon round (`x -> (x + c)^D`).
+///
+/// This function combines two steps for efficiency. It takes a vector of field elements `val`
+/// and adds the pre-formatted round constant `rc`. The result is then raised to the power `D`.
+///
+/// A key optimization is that the round constants `rc` are stored in a negative form `c - P`.
+/// This allows for a simple integer addition `val + rc` without an intermediate modular
+/// reduction. The result of this addition is in the range `[-P, P)`. This non-canonical
+/// result is then reduced back to `[0, P)` before the exponentiation is applied.
+///
+/// # Safety
+/// - `val` must contain elements in canonical form `[0, P)`.
+/// - `rc` must contain elements represented as `c - P`, which are in `[-P, 0]`.
+#[inline(always)]
+fn add_rc_and_sbox<PMP, const D: u64>(val: &mut PackedMontyField31Neon<PMP>, rc: uint32x4_t)
+where
+    PMP: PackedMontyParameters + FieldParameters,
+{
+    unsafe {
+        // Unwrap the field element vector to its raw u32 representation.
+        let vec_val = val.to_vector();
+
+        // Perform a standard wrapping 32-bit addition lane by lane.
+        //
+        // Since `val` is in `[0, P)` and `rc` is in `[-P, 0]` (represented as large u32s),
+        // the wrapping addition correctly computes a result in `[-P, P)`.
+        let val_plus_rc_u32 = aarch64::vaddq_u32(vec_val, rc);
+
+        // Reduce the result from `[-P, P)` back to the canonical range `[0, P)`.
+        // This is required before the exponentiation.
+        //
+        // We use a branchless trick with `vminq` (vector minimum):
+        // - If `val_plus_rc` was positive, it's a small u32. `val_plus_rc + P` is larger,
+        //   so `vminq` returns the original, correct value.
+        // - If `val_plus_rc` was negative, it's a large u32. `val_plus_rc + P` wraps around
+        //   to a small positive u32 (the canonical value), so `vminq` selects this smaller value.
+        let val_plus_rc_plus_p = aarch64::vaddq_u32(val_plus_rc_u32, PMP::PACKED_P);
+        let canonical_val = aarch64::vminq_u32(val_plus_rc_u32, val_plus_rc_plus_p);
+
+        // - Apply the power S-box `x -> x^D`. The input is now correctly in `[0, P)`.
+        // - Update the state.
+        let val_plus_rc_u32_packed = PackedMontyField31Neon::<PMP>::from_vector(canonical_val);
+        *val = val_plus_rc_u32_packed.exp_const_u64::<D>();
```

Comment:
Currently this is doing something equivalent to the simple2 liner:
```suggestion
        *val += rc;
        *val = val.exp_const_u64::<D>();
```

Reply:
To be honest, we can probably just delete this for now and just use `add_rc_and_sbox_generic` until we write our custom function.